Shared/alerts/DynamicEntityLayerAlertTarget.cpp

```diff
@@ -0,0 +1,197 @@
+    info->deleteLater();
+  });
+
+  // subscribe to the observation received signal (this essentially serves as the 'update' signal for the dynamic layer type)
+  connect(m_dynamicEntityLayer->dataSource(), &DynamicEntityDataSource::dynamicEntityObservationReceived, this, [this](DynamicEntityObservationInfo *observationInfo)
+  {
+    // check the entity for a connection and mark the observation as delete later
+    auto dynamicEntity = observationInfo->observation()->dynamicEntity();
+    connectEntityGraphic(dynamicEntity);
+    observationInfo->deleteLater();
+  });
+
+  // subscribe to the purged signal to remove any graphics from the lookup
+  connect(m_dynamicEntityLayer->dataSource(), &DynamicEntityDataSource::dynamicEntityPurged, this, [this](DynamicEntityInfo *info)
+  {
+    // ensure the graphic is actually in the lookup
+    auto dynamicEntity = info->dynamicEntity();
+    if (m_entityGraphics.contains(dynamicEntity->entityId()))
+    {
+      // get a cleanup reference to the graphic that was stored and remove it from the lookup
+      std::unique_ptr<Graphic> graphic(m_entityGraphics[dynamicEntity->entityId()]);
+      m_entityGraphics.remove(dynamicEntity->entityId());
+
+      // signal the data has changed and rebuild the quad tree
+      rebuildQuadtree();
+      emit dataChanged();
+    }
+    info->deleteLater();
+  });
+}
+
+/*!
+  \brief Destructor.
+ */
+Dsa::DynamicEntityLayerAlertTarget::~DynamicEntityLayerAlertTarget()
+{
+}
+
+/*!
+  \brief Returns the list of \l Esri::ArcGISRuntime::Geometry which are in the \a targetArea.
+
+  \note No exact intersection tests are carried out to create this list.
+ */
+QList<Geometry> DynamicEntityLayerAlertTarget::targetGeometries(const Esri::ArcGISRuntime::Envelope &targetArea) const
+{
+  // if the quadtree has been built, use  it to return the set of candidate geometries
+  if (m_quadtree) { return m_quadtree->candidateIntersections(targetArea); }
+
+  // otherwise, return all of the geometry in the overlay
+  QList<Geometry> geometries;
+  for (auto dynamicEntity : m_entityGraphics)
+  {
+    if (!dynamicEntity) { continue; }
+    geometries.append(dynamicEntity->geometry());
+  }
+
+  return geometries;
+}
+
+/*!
+  \brief Returns an empty QVariant.
+ */
+QVariant DynamicEntityLayerAlertTarget::targetValue() const
+{
+  return QVariant{};
+}
+
+/*!
+  \internal
+
+  Connect signals etc. for a new \a dynamicEntity.
+ */
+void DynamicEntityLayerAlertTarget::connectEntityGraphic(Esri::ArcGISRuntime::DynamicEntity *dynamicEntity)
+{
+  // check for an existing graphic
+  Graphic *graphic;
+  if (m_entityGraphics.contains(dynamicEntity->entityId()))
+  {
+    // update the geometry
+    graphic = m_entityGraphics[dynamicEntity->entityId()];
+    graphic->setGeometry(dynamicEntity->geometry());
+  }
+  else
+  {
+      // if no graphic existed for the entity id, construct a new one and insert it into the graphics lookup
+    graphic = new Graphic(dynamicEntity->geometry());
```

Comment:
Who owns this graphic? It either needs a parent, or some other type of memory management. Please review any "new" calls without a parent provided.

Reply:
Wasn't sure what the best practice is for objects like this that are not added to something in the UI like a layer. Graphics created here are deleted via unique_ptr when they are removed from the QMap on the DynamicEntityPurged signal. Should they be given 'this' as their owner in the constructor?

Reply:
I think these should have a parent here. 

However, let me take a step back and ask why you are creating graphics here. You mentioned they aren't meant for the UI. What are they meant for?

Reply:
I think I answered my own question. Graphics are needed for the quadtree collision detection algorithm. Take a look at that code, or other places added graphics to it. There should be a consistent memory management technique in place.

I think there should be a QObject parent provided here, but there may be another pattern in use.

Reply:
Pushed up a new commit where 'this' is provided as the parent in the constructor. This is the case in many other instances elsewhere in the code base.